lib/ecto/query.ex

```diff
@@ -2262,6 +2262,49 @@ defmodule Ecto.Query do
     has_named_binding?(Ecto.Queryable.to_query(queryable), key)
   end
 
+  @doc """
+  Applies the callback function to the query if it does not have a named binding,
+  otherwise returns the query without changes.
+
+  The function accepts an `Ecto.Query` struct and must return the query with an added 
+  named binding.
```

Comment:
I would appreciate it if you have ideas for better wording here and in raised errors.

Reply:
How about this:

```
Applies a callback function to a query if it doesn't contain the given named binding. Otherwise, returns the original query.

The callback function must return an `Ecto.Query` struct that contains the provided named binding, otherwise an error is raised.

This is a convenience function that can be used to simplify conditional modifications of a query. For example, one might write the following code to conditionally add a new named join to a query:

    if has_named_binding?(query, :comments) do
      query
    else
      join(query, :left, c in assoc(p, :comments), as: :comments)
    end

With this function it can be simplified to:

    with_named_binding(query, :comments, fn ->
      join(query, :left, c in assoc(p, :comments), as: :comments)
    end) 

For more information on named bindings see "Named bindings" in this module doc or `has_named_binding/2`. 
``

Reply:
@greg-rychlewski I like your examples, thank you. I'll add some tests to make sure such bindings are supported (did not find a test for that for `as:` and that they can be inspected).